modules/ardrivo/src/Wire.cxx

```diff
@@ -0,0 +1,180 @@
+#include <cstring>
+#include <fmt/format.h>
+#include <range/v3/algorithm/find_if.hpp>
+#include "Arduino.h"
+#include "BoardDataDef.hxx"
+#include "Entrypoint.hxx"
+#include "Error.hxx"
+#include "Wire.h"
+#include "WString.h"
+#include "utility.hxx"
+
+void TwoWire::begin() {
+    begun = true;
+}
+
+void TwoWire::begin(std::uint8_t address) {
+    begin();
+    own_address = address;
+}
+void TwoWire::begin(int sda, int scl) {
+    maybe_init();
+    const auto debug_sig = [=](const char* msg) { return fmt::format("Wire.begin({}, {}): {}", sda, scl, msg); };
+    if (sda >= board_info->pins_caps.size())
+        return handle_error(debug_sig("SDA pin does not exist"));
+    if (scl >= board_info->pins_caps.size())
+        return handle_error(debug_sig("SDL pin does not exist"));
+
+    const auto it = ranges::find_if(board_info->i2c_chans, [=](const I2cChannelInfo& chan){
+        return std::pair{+chan.sda_pin, +chan.scl_pin} == std::pair{sda, scl};
+    });
+    if(it == board_info->i2c_chans.end())
+        return handle_error("SDA SCL pair are not an I2C channel");
+
+    begin();
+    bus_id = std::distance(board_info->i2c_chans.begin(), it);
+}
+void TwoWire::begin(int sda, int scl, std::uint8_t address) {
+    maybe_init();
+    const auto debug_sig = [=](const char* msg) { return fmt::format("Wire.begin({}, {}, {}): {}", sda, scl, +address, msg); };
+    if(address & no_address)
+        handle_error(debug_sig("Invalid address"));
+    begin(sda, scl);
+    begin(address);
+}
+
+std::size_t TwoWire::requestFrom(std::uint8_t address, std::size_t quantity, [[maybe_unused]] bool stop) {
+    if(!begun || quantity == 0)
+        return 0;
+    maybe_init();
+    const auto debug_sig = [=](const char* msg) { return fmt::format("Wire.requestFrom({}, {}, {}): {}", +address, quantity, stop, msg); };
+    if(address & no_address)
+        handle_error(debug_sig("Invalid address"));
+
+    std::unique_lock lock{board_data->i2c_buses[bus_id].devices_mut};
+    auto& device = board_data->i2c_buses[bus_id].devices[address];
+
+    if (device.valueless_by_exception()) // not initialized -- dynamic init
+        return 0; // CAVEAT: [dynamic-init] TWI IO is a no-op at dynamic init
+
+    lock = std::visit(Visitor{
+                   [quantity, lock = std::move(lock)](std::pair<std::condition_variable, std::size_t>& cv_and_len) mutable {
+                        auto& [cv, size] = cv_and_len;
+                       size = quantity;
+                       cv.wait(lock);
```

Comment:
So just for record, when the master calls `requestFrom` the thread has to wait until the slave at least responded with something?

Reply:
Look at what is returned

Reply:
By glancing over the documentation I originally also thought that `requestFrom` should be non-blocking, however since it needs to return how many bytes were effectively transmitted back from the slave, it leaves no choice but to have it blocking.